File: HW7 - Java Implementation of DayofYear and DayofYearSet Classes/DayofYearSet.java
import java.util.ArrayList;
import java.util.Objects;


/**
 * DayofYearSet class keeps DayofYear objects in array
 * lastIndex is shows first empty space in array
 * capacity is total size of the array
 */
public class DayofYearSet {
    private DayofYear[] dateSet; //DayofYear array

    private int lastIndex,capacity;

    /**
     * DayofYear inner class
     */
    public static class DayofYear{
        private int day,month;
        public static int totalDYObjects=0;

        /**
         * No parameter constructor of DayofYear object. It stars the date with 1st January
         */
        public DayofYear() {
            this.day=1;
            this.month=1;
            totalDYObjects++;
        }

        /**
         * This constructor initializes DayofYear class with given day and month
         * If input is invalid, it exits the program with exit code 1
         * @param day is day
         * @param month is month
         */
        public DayofYear(int day, int month) {
            if(month==1 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }
            else if(month==2 && day>0 && day<=28){
                this.day=day;
                this.month=month;
            }
            else if(month==3 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else if(month==4 && day>0 && day<=30){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else if(month==5 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else if(month==6 && day>0 && day<=30){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else if(month==7 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else if(month==8 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }
            else if(month==9 && day>0 && day<=30){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }
            else if(month==10 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }
            else if(month==11 && day>0 && day<=30){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }
            else if(month==12 && day>0 && day<=31){
                this.day=day;
                this.month=month;
                totalDYObjects++;
            }

            else{ //Invalid date case
                System.out.println(day+"."+month+" is invalid Date!");
                System.exit(1);
            }

        }

        public int getDay() {
            return day;
        }

        public int getMonth() {
            return month;
        }

        public static int getTotalDYObjects() {
            return totalDYObjects;
        }


        /**
         * This overridden method combines properties of the class and returns them in String object
         * @return string value of DayofYear class
         */
        @Override
        public String toString() {
            return "DayofYear{" +
                    "day=" + day +
                    ", month=" + month +
                    '}';
        }

        /**
         * This overridden equals method returns true if two objects are sharing the same day and the same month
         * @param obj is Object class(Classic Base class of every class)
         * @return true if day and month values are the same
         */
        @Override
        public boolean equals(Object obj) {
            if (obj == this) {
                return true;
            }
            if (obj == null || obj.getClass() != this.getClass()) {
                return false;
            }

            DayofYear otherClass = (DayofYear) obj;
            return (this.day==otherClass.day && this.month==otherClass.month);
        }




    }


    /**
     * No parameter constructor of DayofYearSet class
     * It initializes lastIndex as 0 and capacity as 10
     */
    public DayofYearSet() {
        this.lastIndex=0;
        this.capacity=10;
        this.dateSet=new DayofYear[capacity];
//        this.dateSet[lastIndex]=new DayofYear(1,1);
//        this.lastIndex++;
    }


    /**
     * Arraylist constructor of DayofYearSet
     * It initializes the capacity as 2 times of size of given arraylist
     * Then it copies all DayofYear Objects of given arraylist
     * @param arrayList containe bunch of DayofYear objects
     */
    public DayofYearSet(ArrayList<DayofYear> arrayList) {
        this.capacity=arrayList.size()*2;
        this.lastIndex=0;
        this.dateSet=new DayofYear[capacity];

        while(lastIndex<arrayList.size()){
            dateSet[lastIndex]=arrayList.get(lastIndex);
            lastIndex++;
        }
        
    }

    /**
     * This constructor is nearly same with default constructor.
     * Only difference is, it creates a new DayofYear object by using given day and month.
     * Then it keeps this object in its own DayofYear[] array
     * @param day is day
     * @param month is month
     */
    public DayofYearSet(int day, int month) {
        this.lastIndex=0;
        this.capacity=10;
        this.dateSet=new DayofYear[capacity];
        this.dateSet[this.lastIndex]=new DayofYear(day,month);
        this.lastIndex++;
    }


    /**
     * @return array of DayofYearSet class
     */
    public DayofYear[] getdateSet() {
        return dateSet;
    }

    /**
     * @return lastIndex of this class
     */
    public int getLastIndex() {
        return lastIndex;
    }

    /**
     * @return lastIndex of this class because it is equal to the size at the same time.
     */
    public int size(){
        return getLastIndex();
    }

    /**
     * @return capacity of the DayofYear class
     */
    public int getCapacity() {
        return capacity;
    }


    /**
     * Creates an empty DayofYear array with doubled capacity.
     * Then it copies all value from initial array of DayofYearSet class
     */
    public void resize(){
        this.capacity*=2;
        DayofYear[] tempDateSet=new DayofYear[capacity];
        for(int i=0;i<this.lastIndex;i++)
            tempDateSet[i]=this.dateSet[i];

        this.dateSet=tempDateSet;
    }

    /**
     * @return Arraylist version of DayofYear[] array of this class
     */
    private ArrayList<DayofYear> arrayToList(){
        ArrayList<DayofYear> arrayList=new ArrayList<DayofYear>();
        for(int i=0;i<this.lastIndex;i++)
            arrayList.add(this.dateSet[i]);
        return arrayList;
    }

    /**
     * It first checks if given DayofYear object exists in the class
     * Then, it checks if lastIndex has reached the capacity(If answer is yes, it calls resize() method
     * Finally it adds given DayofYear object to the lastIndex of DayofYear[] array of this class
     * @param dayofYear is given DayofYear object
     */
    public void add(DayofYear dayofYear){
        if(findIndex(dayofYear)!=-1){
            System.out.println("DayofYear Object: "+dayofYear+" is already in the DayofYearSet class");
            System.out.println("DayofYear object is at index: "+findIndex(dayofYear));
            System.out.println("You can't add a duplicate object!");
            return;
        }

        if(lastIndex==capacity)
            resize();

        this.dateSet[lastIndex]=dayofYear;
        lastIndex++;
    }

    /**
     * This method finds index of given DayofYear object in the DayofYear[] array of this class
     * @param dayofYear is given DayofYear object
     * @return index of the given DayofYear object. It returns -1 if the object does not exist in this class
     */
    private int findIndex(DayofYear dayofYear){
        for(int i=0;i<this.lastIndex;i++){
            if(dayofYear!=null && dateSet[i].equals(dayofYear))
                return i;
        }
        return -1;
    }

    /**
     * This method shifts all DayofYear object to the left by starting from the index of removed object
     * @param dayofYear is given DayofYear object
     */
    public void remove(DayofYear dayofYear){
       if(lastIndex==0){
           System.out.println("You can't remove any element because container is empty!");
           return;
       }

       int removeIndex=findIndex(dayofYear);

       if(removeIndex==-1){
          System.out.println("You can't remove "+dayofYear+" because it doesn't exist in the container");
           return;
       }

       DayofYear d1=dateSet[removeIndex];

       while(removeIndex<lastIndex){
           dateSet[removeIndex]=dateSet[removeIndex+1];
           removeIndex++;
       }
       this.lastIndex--;

       System.out.println(d1+ " has removed from the class");
    }


    /**
     * This method compares all objects of array of this class with another DayofYearSet class' array objects
     * @param obj
     * @return if two classes have the same DayofYear objects in its DayofYear[] array
     */
    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj == null || obj.getClass() != this.getClass()) {
            return false;
        }

        DayofYearSet otherClass = (DayofYearSet) obj;

        for(var dayofyear1: dateSet)
            if(otherClass.findIndex(dayofyear1)==-1 && dayofyear1!=null)
                return false;

        for(var dayofyear2: otherClass.dateSet)
            if(findIndex(dayofyear2)==-1 && dayofyear2!=null)
                return false;

        return true;
    }

    /**
     * @param dayofYearSet
     * @return true if one class have at least one different DayofYear object
     */
    public boolean notEquals(DayofYearSet dayofYearSet){
        return !(this.equals(dayofYearSet));
    }

    /**
     * This class creates an arraylist to store intersection set elements
     * Then It constructs a new DayofYearSet class by giving this arraylist as input parameter
     * @param dayofYearSet
     * @return
     */
    public DayofYearSet intersection(DayofYearSet dayofYearSet){
        ArrayList<DayofYear> dayofYearArrayList=new ArrayList<DayofYear>();
        for(var dayofyear2: dayofYearSet.dateSet)
            if(findIndex(dayofyear2)!=-1)
                dayofYearArrayList.add(dayofyear2);


        dayofYearArrayList.removeIf(Objects::isNull);
        return new DayofYearSet(dayofYearArrayList);
    }

    /**
     * This class creates an arraylist to store difference set elements
     * Then It constructs a new DayofYearSet class by giving this arraylist as input parameter
     * @param dayofYearSet
     * @return
     */
    public DayofYearSet difference(DayofYearSet dayofYearSet){
        ArrayList<DayofYear> dayofYearArrayList=new ArrayList<DayofYear>();
        for(var dayofyear1: dateSet)
            if(dayofYearSet.findIndex(dayofyear1)==-1)
                dayofYearArrayList.add(dayofyear1);

        dayofYearArrayList.removeIf(Objects::isNull);
        return new DayofYearSet(dayofYearArrayList);
    }


    /**
     * This class creates an arraylist to store union set elements
     * Then It constructs a new DayofYearSet class by giving this arraylist as input parameter
     * It combines A/B B/A and AnB
     * @param dayofYearSet
     * @return
     */
    public DayofYearSet union(DayofYearSet dayofYearSet){
        ArrayList<DayofYear> dayofYearArrayList;


        dayofYearArrayList=intersection(dayofYearSet).arrayToList(); //Gets intersection set

        dayofYearArrayList.addAll(difference(dayofYearSet).arrayToList()); //Adds 1st difference set

        dayofYearArrayList.addAll(dayofYearSet.difference(this).arrayToList()); //Adds 2nd difference set

        dayofYearArrayList.removeIf(Objects::isNull);

        return new DayofYearSet(dayofYearArrayList);
    }


    /**
     * This class creates an arraylist to store complement set elements
     * Then It constructs a new DayofYearSet class by giving this arraylist as input parameter
     * @return
     */
    public DayofYearSet complement(){
        ArrayList<DayofYear> dayofYearArrayList=new ArrayList<>();

        DayofYear tempDY;
        int month=1;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=2;
        for(int i=1;i<=28;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=3;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=4;
        for(int i=1;i<=30;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=5;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);

        }

        month=6;
        for(int i=1;i<=30;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);

        }

        month=7;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=8;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=9;
        for(int i=1;i<=30;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=10;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=11;
        for(int i=1;i<=30;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }

        month=12;
        for(int i=1;i<=31;i++){
            tempDY=new DayofYear(i,month);
            if(findIndex(tempDY)==-1)
                dayofYearArrayList.add(tempDY);
        }


        dayofYearArrayList.removeIf(Objects::isNull);

        return new DayofYearSet(dayofYearArrayList);
    }


    /**
     * This overridden toString method returns size,lastIndex and array elements of DayofYearSet class
     * @return properties of this class as string value
     */
    @Override
    public String toString() {
        String properties= "DayofYearSet{ " +
                "lastIndex=" + lastIndex +
                ",\t capacity=" + capacity +
                ",\ndateSet=[\n";
        for(int i=0;i<lastIndex;i++)
            if(dateSet[i]!=null)
                properties = properties + (dateSet[i] + ",\n");

        properties=properties+"]\n}";
        return properties;
    }
}

File: HW4 - DayofYear and DayofYearSet Classes/DayofYearSet.cpp
//
// Created by baris on 2.12.2021.
//

#include "DayofYearSet.h"


using namespace std;

/**
 * Initializing static integer totalDayofYearObjects
 */
int baris::DayofYearSet::DayofYear::totalDayofYearObjects=0;

/**
 * This constructor first checks given month and combines it
 * with the given day.
 * It exits if the date is illegal
 * @param day is day
 * @param month is month
 */
baris::DayofYearSet::DayofYear::DayofYear(int day, int month) {

    if(month==1 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }
    else if(month==2 && day>0 && day<=28){
        this->day=day;
        this->month=month;
    }
    else if(month==3 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }

    else if(month==4 && day>0 && day<=30){
        this->day=day;
        this->month=month;
    }

    else if(month==5 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }

    else if(month==6 && day>0 && day<=30){
        this->day=day;
        this->month=month;
    }

    else if(month==7 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }

    else if(month==8 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }
    else if(month==9 && day>0 && day<=30){
        this->day=day;
        this->month=month;
    }
    else if(month==10 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }
    else if(month==11 && day>0 && day<=30){
        this->day=day;
        this->month=month;
    }
    else if(month==12 && day>0 && day<=31){
        this->day=day;
        this->month=month;
    }

    else{
        cerr<<day<<"."<<month<<" is invalid date!"<<endl;
        exit(1);
    }

    totalDayofYearObjects++;
}
/**
 * This no parameter contructor creates DayofYear object with 01/01 date
 */
baris::DayofYearSet::DayofYear::DayofYear() {
    this->day=1;
    this->month=1;

    totalDayofYearObjects++;

}
/**
 * Desctructor of DayofYear object
 */
baris::DayofYearSet::DayofYear::~DayofYear() {

}

int baris::DayofYearSet::DayofYear::getDay() const {
    return day;
}



int baris::DayofYearSet::DayofYear::getMonth() const {
    return month;
}

int baris::DayofYearSet::DayofYear::getTotalDayofYearObjects() {
    return totalDayofYearObjects;
}


/**
 * This method compares two DayofYear classes
 * @param d2 DayofYear object
 * @return true if two objects have the same day and month
 */
bool baris::DayofYearSet::DayofYear::operator==(const baris::DayofYearSet::DayofYear &d2) const {
    return(getDay()==d2.getDay() && getMonth()==d2.getMonth());
}

/**
 * Default constructor of DayofYearSet class
 * It adds 01/01 date to vector of dateSet
 */
baris::DayofYearSet::DayofYearSet() {
    this->dateSet.push_back(new DayofYear(1,1));
}

/**
 * This constructor initialize the class witj given day and month
 * @param day
 * @param month
 */
baris::DayofYearSet::DayofYearSet(int day, int month) {
    this->dateSet.push_back(new DayofYear(day,month));
}
/**
 * This constructor constructs the class by given DayofYear* vector
 * @param dateSet
 */
baris::DayofYearSet::DayofYearSet(const vector<DayofYear *> &dateSet) : dateSet(dateSet) {}

/**
 * Getter method of DayofYear* vector that named dateSet
 * @return
 */
const vector<baris::DayofYearSet::DayofYear *> &baris::DayofYearSet::getdateSet() const {
    return dateSet;
}

/**
 * This method searches a DayofYear object in vector
 * by given DayofYear* pointer
 * @param d1
 * @return true if it finds the object in the vector of dateSet
 */
bool baris::DayofYearSet::search(baris::DayofYearSet::DayofYear* d1) const{
    for(DayofYear* d2 : dateSet)
        if(*d1==*d2)
            return true;
    return false;
}

/**
 * This method finds the index of specifix DayofYear* pointer
 * by given DayofYear* pointer
 * @param d1
 * @return index of given element in the vector
 */

int baris::DayofYearSet::findIndex(baris::DayofYearSet::DayofYear* d1) const {
    for(int i=0;i<dateSet.size();i++)
        if(*d1==*dateSet[i])
            return i;
    return -1;
}

/**
 * This method adds a new DayofYear* pointer to the dateSet vector
 * If it detects a duplicate element, it warns user and doesn't add anything
 * to dateSet vector
 * @param day
 * @param month
 */
void baris::DayofYearSet::add(int day, int month) {
    DayofYear* d1=new DayofYear(day,month);
    if(search(d1))
        cout<<d1->getDay()<<"."<<d1->getMonth()<<" is already in the set."
        <<"You can't add a duplicate one!"<<endl;
    else{
        cout<<d1->getDay()<<"."<<d1->getMonth()<<" is added to the DayofYearSet"<<endl;
        dateSet.push_back(d1);
    }
}
/**
 * It removes a DayofYear* pointer by given day and month
 * @param day
 * @param month
 */
void baris::DayofYearSet::remove(int day,int month) {
    DayofYear* d1=new DayofYear(day,month);
    if(search(d1)){
        int elementIndex= findIndex(d1);
        DayofYear* realD1=dateSet[elementIndex];
        dateSet.erase(dateSet.begin()+elementIndex);
        delete realD1;
        realD1=nullptr;
        cout<<d1->getDay()<<"."<<d1->getMonth()<<
        " is removed from DayofYearSet"<<endl;
    }
    else{
        cout<<d1->getDay()<<"."<<d1->getMonth()<<" does not exist in DayofYearSet Class!"<<endl;
    }
    delete d1;
    d1= nullptr;

}

/**
 * This method returns size of dateSet vector
 * @return
 */

int baris::DayofYearSet::size() const {
    return dateSet.size();
}

/**
 * This overloaded operator == searches every element of given class in dateSet vector
 * It repeats operation by converse and returns false if it detects any different element
 * @param ds2
 * @return
 */
bool baris::DayofYearSet::operator==(const baris::DayofYearSet &ds2) const {
    for(DayofYear* d1: dateSet){
        if(ds2.search(d1)==false)
            return false;
    }

    for(DayofYear* d2: ds2.getdateSet()){
        if(search(d2)==false) //double controlling to prevent subset cases
            return false;
    }
    return true;
}
/**
 * This overloaded operator != returns just inverse version of == operator
 * @param ds2
 * @return
 */
bool baris::DayofYearSet::operator!=(const baris::DayofYearSet &ds2) const {
    return !(*this==ds2);
}


/**
 * This method removes duplicate elements in dateSet vector
 * It is called in + operator overloading to fix union set
 * @param daysVector
 */
void baris::DayofYearSet::removeDuplicates(vector<DayofYear*>& daysVector) const{
    for(int i=0;i<daysVector.size();i++)
        for(int j=i+1;j<daysVector.size();j++)
            if(*daysVector[i]==*daysVector[j]){
                DayofYear* duplicate=daysVector[j];
                delete duplicate;
                duplicate= nullptr;
                daysVector.erase(daysVector.begin()+j);
            }
}
/**
 * This overloaded operator + creates a new DayofYear* vector
 * It combines DayofYear elements of this class and given class in the vector
 * It removes duplicate elements in the vector and finally returns a new DayofYearSet class by using this vector
 * @param ds2
 * @return
 */
const baris::DayofYearSet baris::DayofYearSet::operator+(const baris::DayofYearSet &ds2) const {
    vector<DayofYear*> unionVector;
    for(DayofYear* d1: dateSet)
        unionVector.push_back(new DayofYear(d1->getDay(),d1->getMonth()));

    for(DayofYear* d2: ds2.getdateSet())
        unionVector.push_back(new DayofYear(d2->getDay(),d2->getMonth()));

    removeDuplicates(unionVector);

    return DayofYearSet(unionVector);
}

/**
 * This overloaded operator - creates a new DayofYear* vector
 * Then it searches all elements of this class in given class
 * Then it stores different elements in the vector and returns a new class by using this vector
 * @param ds2
 * @return
 */
const baris::DayofYearSet baris::DayofYearSet::operator-(const baris::DayofYearSet &ds2) const {
    vector<DayofYear*> differenceSet;
    for(DayofYear* d1: dateSet)
        if(!ds2.search(d1))
            differenceSet.push_back(d1);
    
    return DayofYearSet(differenceSet);    
}
/**
 * This overloaded operator ^ creates a new DayofYear* vector
 * It searches elements of given class in this class' vector
 * It stores common elements in the vector and returns a new class by using this vector
 * @param ds2
 * @return
 */
const baris::DayofYearSet baris::DayofYearSet::operator^(const baris::DayofYearSet &ds2) const {
    vector<DayofYear*> intersectionSet;
    for(DayofYear* d2: ds2.getdateSet())
        if(search(d2))
            intersectionSet.push_back(d2);

    return DayofYearSet(intersectionSet);
}


/**
 * This overloaded operator ! creates a new DayofYear* vector to store elements
 * It iterates through all 365 days in year
 * It calls search method for all 365 days and stores elements if search method returns false
 * Finally it returns a new DayofYearSet object by using this vector
 * @return
 */
const baris::DayofYearSet baris::DayofYearSet::operator!() const {
    vector<DayofYear*> complementSet;
    DayofYear* tempPointer;
    int month=1;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=2;
    for(int i=1;i<=28;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=3;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
       // delete tempPointer;
        //tempPointer= nullptr;
    }

    month=4;
    for(int i=1;i<=30;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=5;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=6;
    for(int i=1;i<=30;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=7;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=8;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=9;
    for(int i=1;i<=30;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
       // delete tempPointer;
        //tempPointer= nullptr;
    }

    month=10;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=11;
    for(int i=1;i<=30;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    month=12;
    for(int i=1;i<=31;i++){
        tempPointer=new DayofYear(i,month);
        if(!search(tempPointer))
            complementSet.push_back(tempPointer);
        //delete tempPointer;
        //tempPointer= nullptr;
    }

    //delete tempPointer;
    //tempPointer= nullptr;

    removeDuplicates(complementSet);

    return DayofYearSet(complementSet);
}
/**
 * This method returns i th element of dateSet vector
 * @param index
 * @return
 */
baris::DayofYearSet::DayofYear *baris::DayofYearSet::operator[](int index) {
    return dateSet[index];
}










